psydac/core/bsplines.py

```diff
@@ -801,32 +812,58 @@ def basis_ders_on_quad_grid(knots, degree, quad_grid, nders, normalization):
 
     Returns
     -------
-    basis: 4D numpy.ndarray
+    basis: ndarray
         Values of B-Splines and their derivatives at quadrature points in
         each element of 1D domain. Indices are
         . ie: global element         (0 <= ie <  ne    )
         . il: local basis function   (0 <= il <= degree)
         . id: derivative             (0 <= id <= nders )
         . iq: local quadrature point (0 <= iq <  nq    )
 
-    """
-    # TODO: add example to docstring
+    Examples
+    --------
+    >>> knots = np.array([0.0, 0.0, 0.25, 0.5, 0.75, 1., 1.])
+    >>> degree = 2
+    >>> bk = breakpoints(knots, degree)
+    >>> # Valid grid
+    >>> grid = np.array([np.linspace(bk[i], bk[i+1], 4, endpoint=False) for i in range(len(bk) - 1)])
+    >>> # Also a valid grid (doesn't start on the left boundary of the domain)
+    >>> grid_2 = grid[2:, :]
+    >>> basis_ders_on_quad_grid(knots, degree, grid, 0, "B")
+    array([[[[0.5, 0.28125, 0.125, 0.03125]],
+            [[0.5, 0.6875 , 0.75 , 0.6875 ]],
+            [[0. , 0.03125, 0.125, 0.28125]]],
+           [[[0.5, 0.28125, 0.125, 0.03125]],
+            [[0.5, 0.6875 , 0.75 , 0.6875 ]],
+            [[0. , 0.03125, 0.125, 0.28125]]]])
+    >>> basis_ders_on_quad_grid(knots, degree, grid_2, 0, "B")
+    array([[[[0.5, 0.28125, 0.125, 0.03125]],
+            [[0.5, 0.6875 , 0.75 , 0.6875 ]],
+            [[0. , 0.03125, 0.125, 0.28125]]]])
 
-    ne,nq = quad_grid.shape
+    """
+    ne, nq = quad_grid.shape
     basis = np.zeros((ne, degree+1, nders+1, nq))
 
     if normalization == 'M':
         scaling = 1. / basis_integrals(knots, degree)
+    spans = elements_spans(knots, degree)
+
+    assert ne <= len(spans)
+    # Test to see if the grid doesn't start at 0
+    offset = 0
+    span_exact = find_span(knots, degree, quad_grid[0, 0])
+    if span_exact != spans[0]:
+        offset = span_exact - spans[0]
```

Comment:
I am not sure of what you are trying to achieve here. If the user wants to use an arbitrary quadrature grid, this function cannot handle it. So we need to specify the exact assumptions in the docstring.

Does the current version of this function handle the special case in `psydac/api/fem.py`? It seems to me that it should, thanks to how you compute the `offset` variable.

Reply:
I was trying to make it possible to use grid that are shifted like in the example I added to the docstring. This special case is not used anywhere in the code currently. It can handle the special boundary case.